### ISSUE TITLE
Force hooks to execute even when no cloudformation change

### ISSUE THREAD

**Issue by ghost** [Sceptre/sceptre#219]
Hi

Currently hooks appear only to execute when there is a change to an underlying template. We have a number of use cases where we use hooks to configure things both inside the AWS ecosytem (where there is no or limited cloudformation support) or outside the AWS ecosystem. an example includes setting up federated SSO by creating a SAML provider (via hook & boto3), creating some roles and policies with cloudformation and also then configuring groups in an external AD system. Another example includes setting up roles for cloudhealth to access account data in cloudformation and then registering the account into cloudhealth from a hook calling a web api. 

As a result there are situations where we make changes to hooks but not to the cloudformation template. However we then have to make some sort of minor change to the cloudformation template to get the hook to execute. 

Therefore I would like for there to be a way to force hooks to execute even when there is no cloudformation change. This would make some of our changes a lot simpler.

Thanks
Rob Godfrey

**Comment by theseanything** [comment]
Similar to #208

**Comment by b-t-g** [comment]
Listing the use-case given in #208 (by @cboss24) for ease of reference:
Zipping and uploading code to be executed in AWS Lambda, even when there's no Cloudformation change.

**Comment by theseanything** [comment]
For this we'd need to implement some sort of flag within the config - maybe like this:
```yaml
hooks:
    after_create: 
        - !cmd "echo hello world"
    after_update: 
        - run: always
          hook: !cmd "echo hello earth"
        - run: conditional
          hook: !cmd "echo hello venus"
```
whereby a `run` property is specified at the command level and with value of "always" or "conditional".

**Comment by b-t-g** [comment]
I really like the format of that command, here are some thoughts for the exact syntax and semantics:

One may want to run an "after_update" hook only after a cloudformation change has been made, only when there is no cloudformation change made, and after every update, so I think the value corresponding to "run" should make this trichotomy explicit.

Now we just have to think of something that makes sense (and ideally makes sense even without having to read this github issue for context).

**Comment by b-t-g** [comment]
A further idea for naming:

When someone only wants something to explicitly run something after a cloudformation stack has been updated or launched, it'll probably affect or use the stack that they created. Not necessarily, but most of the time.

When someone wants to explicitly run something when no cloudformation stack has been updated  or launch, it is likely independent of the stack.

So my proposal for values of the run key is:
- always: Always run the command/hook.
- stack_dependent: Only run the command/hook if a stack has been successfully launched or updated.
- stack_independent: Only run the command/hook if a stack was not launched or update.

**Comment by theseanything** [comment]
I agree for the trichotomy of values - I feel `stack_dependent` and `stack_independent` should be more succinct.

**Comment by b-t-g** [comment]
I've slowly convinced myself that it's not a huge deal; given the primary use-case of Sceptre, it seems like ```stack_independent``` and ```always``` are somewhat rare exceptions rather than the rule and I would be surprised if there was a shorter name that was just as descriptive. How about changing ```stack_dependent``` to ```default``` and leaving ```stack_independent``` as it is?

**Comment by theseanything** [comment]
Can we open a separate issue to discuss hook properties? and close this one.

**Comment by cboss24** [comment]
@theseanything @b-t-g, anything I can help with to move this forward?

**Comment by b-t-g** [comment]
@cboss24, I've started working on this in the referenced commits from earlier, but I haven't had too much time to dedicate to Sceptre recently. It looks like the branch I committed the changes to were deleted when the 2.0 branch became the de-facto master branch. If you want to take it over from here, I would be delighted; I think holding onto it any longer would definitely be treading "cookie-licking" territory.

**Comment by cboss24** [comment]
Okay! I am happy to look at it. @seanrankine, you mentioned wanting to open a new issue to discuss hooks. Is there anything specific you'd like to discuss before I start working on this?

**Comment by b-t-g** [comment]
@cboss24, I just talked to Sean and we think the "further discussion" probably had something to do with how to name the functionality. Naturally, there's the "always" execute hook, but what should we call the current functionality and, is the case that I called "stack_independent" above worth considering (I'd still say yes) and, if so, what should we call it? If you're okay with possibly renaming stuff in your PR, there's no reason for you to not go for it.

**Comment by b-t-g** [comment]
Continuing on my "Don't over-optimize all the things" from another issue; maybe we should abandon what I call the "stack independent" case above? While I can think of some use cases for it, they may be a bit contrived and could possibly never be used. We can always just add it if we need to.

**Comment by cboss24** [comment]
Okay, that simplifies the solution a little bit. With that in mind, what do you think about adding a new hook_points to support this. Something like `always_after_create`, `always_after_update`, `always_after_delete`. This way we don't have to change the hook declaration from a scalar.

**Comment by b-t-g** [comment]
I've thought about it a bit and it looks like there are two options, your idea and to add a field to that "hook block" instead; i.e.,:

```yaml
hooks:
  before_create:
    always: true
    cmd: <whatever>
  before_create:
    cmd: <Something we don't want to happen if not created>
```

If your idea is relatively simple to implement, (i.e., it doesn't require having modify every existing hook), then I'm all for it!

**Comment by b-t-g** [comment]
I had, until now been forgetting @theseanything's suggestion above, which I also like:

```yaml
hooks:
    after_create: 
        - !cmd "echo hello world"
    after_update: 
        - run: always
          hook: !cmd "echo hello earth"
        - run: conditional
          hook: !cmd "echo hello venus"
```

But it could be a bit verbose for having many "always hooks" and I think my original implementation used this syntax and there a few wonky checks to see whether there was a run command. That wasn't a huge deal when that kind of flexibility was great for handling three cases but, now that we're now focusing on two cases, maybe that flexibility isn't necessary?

After thinking about it a bit more, I like your idea more than the one I proposed since it is more explicit, easier to determine what it does just by looking at it, and thus more intuitive. And, as an added bonus, yours is more succinct than mine.

**Comment by zaro0508** [comment]
This is related to issue #279
sceptre-date-resolver has been implemented in issue #608 which may help with this issue.
